### ISSUE TITLE
conda run uses a temp file in the environment directory

### ISSUE THREAD

**Issue by hjoliver** [conda/conda#10098]
## Current Behavior

`conda run` writes an activation script to a temporary file _in the target environment directory_, which means it can't be used with shared environments (unless all users have write access to the conda location - not a good idea)

### Steps to Reproduce

```
$ export CONDA_TEST_SAVE_TEMPS=true
$ conda run -n drugs_and_money cylc --version
WARNING conda.cli.main_run:execute(39): CONDA_TEST_SAVE_TEMPS :: retaining main_run script_caller \
    /home/oliverh/miniconda3/envs/drugs_and_money/.tmprsaedgg3
8.0a2

$ cat /home/oliverh/miniconda3/envs/drugs_and_money/.tmprsaedgg3
eval "$(/home/oliverh/miniconda3/bin/conda shell.posix hook)"
conda activate  /home/oliverh/miniconda3/envs/drugs_and_money
cylc --version
```

## Expected Behavior

`conda run` should write the temporary activation script to a standard temp file location for the user.

## Environment Information
<details open><summary><code>`conda info`</code></summary><p>

```
   active environment : /tmp
    active env location : /tmp
            shell level : 0
       user config file : /home/oliverh/.condarc
 populated config files : /home/oliverh/.condarc
          conda version : 4.8.3
    conda-build version : not installed
         python version : 3.7.7.final.0
       virtual packages : __glibc=2.17
       base environment : /home/oliverh/miniconda3  (writable)
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /home/oliverh/miniconda3/pkgs
                          /home/oliverh/.conda/pkgs
       envs directories : /home/oliverh/miniconda3/envs
                          /home/oliverh/.conda/envs
               platform : linux-64
             user-agent : conda/4.8.3 requests/2.24.0 CPython/3.7.7 Linux/3.10.0-1127.13.1.el7.x86_64 centos/7.8.2003 glibc/2.17
                UID:GID : 8560:10000
             netrc file : None
           offline mode : False
```

</p></details>

<details open><summary><code>`conda config --show-sources`</code></summary><p>

```
auto_activate_base: False
ssl_verify: True
channel_priority: disabled
channels:
  - conda-forge
unsatisfiable_hints: True
anaconda_upload: False
```

</p></details>

<details><summary><code>`conda list --show-channel-urls`</code></summary><p>

```
conda list --show-channel-urls
# packages in environment at /home/oliverh/miniconda3:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main    defaults
brotlipy                  0.7.0           py37h7b6447c_1000    defaults
ca-certificates           2020.6.24                     0    defaults
certifi                   2020.6.20                py37_0    defaults
cffi                      1.14.0           py37he30daa8_1    defaults
chardet                   3.0.4                 py37_1003    defaults
conda                     4.8.3                    py37_0    defaults
conda-package-handling    1.6.0            py37h8f50634_2    conda-forge
cryptography              2.9.2            py37h1ba5d50_0    defaults
idna                      2.10                       py_0    defaults
ld_impl_linux-64          2.34                 h53a641e_7    conda-forge
libedit                   3.1.20191231         h14c3975_1    defaults
libffi                    3.3                  he6710b0_2    defaults
libgcc-ng                 9.2.0                hdf63c60_0    conda-forge
libstdcxx-ng              9.2.0                hdf63c60_2    conda-forge
ncurses                   6.2                  he6710b0_1    defaults
openssl                   1.1.1g               h7b6447c_0    defaults
pip                       20.1.1                   py37_1    defaults
pycosat                   0.6.3            py37h7b6447c_0    defaults
pycparser                 2.20                       py_2    defaults
pyopenssl                 19.1.0                     py_1    defaults
pysocks                   1.7.1                    py37_1    defaults
python                    3.7.7                hcff3b4d_5    defaults
python_abi                3.7                     1_cp37m    conda-forge
readline                  8.0                  h7b6447c_0    defaults
requests                  2.24.0                     py_0    defaults
ruamel_yaml               0.15.87          py37h7b6447c_1    defaults
setuptools                49.2.0                   py37_0    defaults
six                       1.15.0                     py_0    defaults
sqlite                    3.32.3               h62c20be_0    defaults
tk                        8.6.10               hbc83047_0    defaults
tqdm                      4.48.0             pyh9f0ad1d_0    conda-forge
urllib3                   1.25.9                     py_0    defaults
wheel                     0.34.2                   py37_0    defaults
xz                        5.2.5                h7b6447c_0    defaults
yaml                      0.2.5                h7b6447c_0    defaults
zlib                      1.2.11               h7b6447c_3    defaults
```

</p></details>

┆Issue is synchronized with this [Jira Task](https://anaconda.atlassian.net/browse/CONDAORG-124)

**Comment by hjoliver** [comment]
```python
# conda/cli/main_run.py:21
    prefix = context.target_prefix or os.getenv("CONDA_PREFIX") or context.root_prefix
```
Where `contex.target_prefix` appears to be the target environment location.

**Comment by tseemann** [comment]
We are also having this rpoblem.

We want to use `conda run -p /;paht/to/env` where that folder is a env owned by abother user.

It fails because `conda run` creates a temp file in `--prefix` which other ysers do not have permissions.

It should be honouring the POSIX `$TMPDIR` setting.

**Comment by claresloggett** [comment]
The line of code mentioned by @hjoliver is https://github.com/conda/conda/blob/729e8b2ccb1031dc1f1e5384ff0806b20b6d6fa7/conda/cli/main_run.py#L22

This does appear to be the problem - is there any reason the tempfile needs to be in this subdirectory? Would there be an issue with creating it the standard ($TMPDIR) way?

**Comment by jonashaag** [comment]
Just hit this too, any suggested workarounds?

**Comment by jonashaag** [comment]
Aha, this seems to fixed with https://github.com/conda/conda/pull/11215?

**Comment by hjoliver** [comment]
Closed by #11215 :tada:

**Comment by tseemann** [comment]
So happy to see this finally resolved!
Thank you all!
